protocols/dcutr/src/behaviour_impl.rs

```diff
@@ -282,32 +274,24 @@ impl NetworkBehaviour for Behaviour {
         addr: &Multiaddr,
         role_override: Endpoint,
     ) -> Result<THandler<Self>, ConnectionDenied> {
-        match self
-            .outgoing_direct_connection_attempts
-            .remove(&(connection_id, peer))
+        if is_relayed(addr) {
+            return Ok(Either::Left(handler::relayed::Handler::new(
+                ConnectedPoint::Dialer {
+                    address: addr.clone(),
+                    role_override,
+                },
+            ))); // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
+        }
+
+        if let Some(&relayed_connection_id) = self.direct_to_relayed_connections.get(&connection_id)
         {
-            None => {
-                let handler = if is_relayed(addr) {
-                    Either::Left(handler::relayed::Handler::new(ConnectedPoint::Dialer {
-                        address: addr.clone(),
-                        role_override,
-                    })) // TODO: We could make two `handler::relayed::Handler` here, one inbound one outbound.
-                } else {
-                    Either::Right(Either::Right(dummy::ConnectionHandler))
-                };
-
-                Ok(handler)
-            }
-            Some(_) => {
-                assert!(
-                    !is_relayed(addr),
-                    "`Prototype::DirectConnection` is never created for relayed connection."
-                );
-
-                Ok(Either::Right(Either::Left(
-                    handler::direct::Handler::default(),
-                )))
-            }
+            self.outgoing_direct_connection_attempts
+                .remove(&(relayed_connection_id, peer));
```

Comment:
Can we have an `assert!` here that this removed an entry?

If this doesn't remove an entry, we have a bug somewhere in our state tracking.

Reply:
`outgoing_direct_connection_attempts` will not contain an entry on the dialer's end, which is why the `dcutr` CI test was failing.

Reply:
Ah yes, we need to check `role_override` here. If that is set to `Endpoint::Listener`, we MUST have an entry in this hashmap.

Reply:
I've added the assert for listener.